### PR TITLE
improve notification routing doc 2

### PR DESCRIPTION
**Pull request by jessereynolds**
Same as #21 but with some additional commits.

### CODE REVIEW

**Comment by jessereynolds** [comment]
@ali-graham and @Hobbsee your feedback on this would be appreciated

**Comment by ali-graham** [comment]
Hmm, I hadn't seen Github's image diff-er before... nice!

Yup, looks good.

File: source/docs/1.0/usage/Notification-Routing.md [hunk 1]
```diff
@@ -1,15 +1,57 @@
...
 ![notification routing](/images/notification-routing.gif)
 
-#### Creating the above animation
+## Details
 
-1. Export [architecture diagrams](/images/FlapjackArchitecture.key) to PNG
-2. Run this shell:
+1. **Receive events.**
+    * Events are generated by external check execution systems and created as json objects in the `events` queue in Redis.
+    * Processor removes events off the queue and, astonishingly, processes them one by one.
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: The tense on this line doesn't match the rest of the document, and astonishingly probably isn't needed.

File: source/docs/1.0/usage/Notification-Routing.md [hunk 2]
```diff
@@ -1,15 +1,57 @@
...
+
+1. **Determine when to generate a notification**. The event is discarded by the filters if ANY of the following are true:
+     * check is OK and it's the first event we've seen for this check (new check)
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: Should be: check state is Ok

File: source/docs/1.0/usage/Notification-Routing.md [hunk 3]
```diff
@@ -1,15 +1,57 @@
...
+     * check is OK and previous state was OK
+     * check is in scheduled or unscheduled maintenance
+     * check is failing and duration of current failure is less than the `initial_failure_delay` (default 30 seconds, can be specified in the incoming event)
+     * check is failing and time elapsed since last notification is less than `repeat_failure_delay` (default 60 sec, can be specified in the incoming event)
+
+2. **Find interested contacts**
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: Interested is the wrong word here.  Linked?  Relevant?

File: source/docs/1.0/usage/Notification-Routing.md [hunk 4]
```diff
@@ -1,15 +1,57 @@
...
+     * [contacts](/docs/1.0/jsonapi#contacts) that have this event's [entity](/docs/1.0/jsonapi#entities)
+     * contacts that have the special entity with id
+       [ALL](/docs/1.0/usage/Howto-Dynamic-Entity-Contact-Linking)
+
+3. **For each contact**
+  3. **Find the subset of applicable notification rules based on entity, tags, severity,
+     time of day**. Find all the
+     [notification rules](/docs/1.0/jsonapi#notification-rules) for the contact,
+     for which **all** of the following are true:
+      * If the rule has any `tags` then they must all be present in the event's tags. Note that
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: comma after 'tags'.  Same goes for the next bullet points.

Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: Are the tags in the notification rules listed by value, or a key-value pair?  If the latter, are the tags automatically created by the processor the exact ones below, or do they have slightly different names?

File: source/docs/1.0/usage/Notification-Routing.md [hunk 5]
```diff
@@ -1,15 +1,57 @@
...
+        an event has a number of tags created automatically by the processor: the entity name, domain, hostname, and each of the words in the check name. Tags can also be added to incoming events by check execution systems, eg [flapjackfeeder](https://github.com/flapjack/flapjackfeeder).
+      * If the rule has any `regex_tags` then they must all match at least one of the event's tags.
+      * If the rule has any `entities` then one of them must be the event's entity.
+      * If the rule has any `regex_entities` then at least one of them must match the event's entity.
+      * If the rule has any `time_restrictions` then at least one of them must match the event's timestamp.
+
+  4. **Skip contact based on blackhole notification rules**.
+      * If any of the remaining notification rules are blackhole rules for the event's state (eg
+         `warning_blackhole == true` when event's `state == warning`) we do not notify
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: comma after )

Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: In fact, I'd move the example to the end of this sentence.

File: source/docs/1.0/usage/Notification-Routing.md [hunk 6]
```diff
@@ -1,15 +1,57 @@
...
-for f in *.png ; do convert $f $(basename $f .png).gif ; done
-gifsicle --delay=200 --loop *.gif > anim.gif
-```
+  5. **Find the notifiable media for the contact based on notification rules**.
+      * From the remaining notification rules, get the
+          media listed in each rule for the event's
+         state (eg if event's `state == warning` then get all `warning_media` from the remaining notification rules).
+      * If there are no notifiable media, we do not notify the current contact.
+
+  6. **Discard media based notification interval**.
```
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: media-based

File: source/docs/1.0/usage/Notification-Routing.md (hunk 3 above)
Comment (jessereynolds) [inline source/docs/1.0/usage/Notification-Routing.md]: Are you suggesting that alert recipients are a disinterested lot? We use 'interested' a fair bit. It's in the gif, for eg.

File: source/docs/1.0/usage/Notification-Routing.md (hunk 6 above)
Comment (Hobbsee) [inline source/docs/1.0/usage/Notification-Routing.md]: This doesn't mean anything to me without reading the text below - is there a better name for this?  Perhaps "Filter on media-based frequency threshold".

File: source/docs/1.0/usage/Notification-Routing.md (hunk 4 above)
Comment (jessereynolds) [inline source/docs/1.0/usage/Notification-Routing.md]: value, no structure within the value is enforced

**Comment by Hobbsee** [comment]
That's a really nice gif, and I learned some stuff around this today.  Thanks, @necula01 & @jessereynolds!

File: source/docs/1.0/usage/Notification-Routing.md (hunk 5 above)
Comment (jessereynolds) [inline source/docs/1.0/usage/Notification-Routing.md]: really? that seems wrong to me

File: source/docs/1.0/usage/Notification-Routing.md (hunk 1 above)
Comment (jessereynolds) [inline source/docs/1.0/usage/Notification-Routing.md]: it's current tense, like the rest of the document. am I missing something?

**Pull request merged**